File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/structure/TemplateInstanceExecution.java
package net.nemerosa.ontrack.model.structure;

import lombok.Data;

import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * Execution context for the creation of a template instance.
 */
@Data
public class TemplateInstanceExecution {

    /**
     * Replacement function to use
     */
    private final Function<String, String> replacementFn;

    /**
     * List of resolved template definition parameters
     */
    private final Map<String, String> parameterValues;

    /**
     * Performs a replacement
     */
    public String replace(String value) {
        return replacementFn.apply(value);
    }

    public List<TemplateParameterValue> asTemplateParameterValues() {
        return parameterValues.entrySet().stream()
                .map(entry -> new TemplateParameterValue(entry.getKey(), entry.getValue()))
                .collect(Collectors.toList());
    }
}

File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/structure/TemplateInstanceExecution.kt
package net.nemerosa.ontrack.model.structure

/**
 * Execution context for the creation of a template instance.
 *
 * @property replacementFn Replacement function to use
 * @property parameterValues List of resolved template definition parameters
 */
class TemplateInstanceExecution(
        val replacementFn: (String) -> String,
        val parameterValues: Map<String, String>
) {

    /**
     * Performs a replacement
     */
    fun replace(value: String): String {
        return replacementFn(value)
    }

    fun asTemplateParameterValues(): List<TemplateParameterValue> {
        return parameterValues.map { (key, value) ->
            TemplateParameterValue(key, value)
        }
    }
}
